indexer/src/store.rs

```diff
@@ -0,0 +1,1240 @@
+                    delete_buffer: HashSet::new(),
+                };
+                process(&mut batch);
+                batch.commit();
+            }
+            Err(err) => {
+                error!(target: "indexer", "indexer db failed to create new batch, error: {:?}", err);
+            }
+        }
+    }
+
+    pub fn sync_index_states(&self) {
+        info!(target: "indexer", "Start sync index states with chain store");
+        let mut lock_hash_index_states = self.get_lock_hash_index_states();
+        if lock_hash_index_states.is_empty() {
+            return;
+        }
+
+        // retains the lock hashes on fork chain and detach blocks
+        lock_hash_index_states.retain(|_, index_state| {
+            self.shared
+                .store()
+                .get_block_number(&index_state.block_hash)
+                != Some(index_state.block_number)
+        });
+        lock_hash_index_states
+            .iter()
+            .for_each(|(lock_hash, index_state)| {
+                let mut index_lock_hashes = HashSet::new();
+                index_lock_hashes.insert(lock_hash.to_owned());
+
+                let mut block = self
+                    .shared
+                    .store()
+                    .get_block(&index_state.block_hash)
+                    .expect("block exists");
+                // detach blocks until reach a block on main chain
+                self.commit_batch(|batch| {
+                    self.detach_block(batch, &index_lock_hashes, &block);
+                    while self
+                        .shared
+                        .store()
+                        .get_block_hash(block.header().number() - 1)
+                        != Some(block.header().parent_hash().to_owned())
+                    {
+                        block = self
+                            .shared
+                            .store()
+                            .get_block(block.header().parent_hash())
+                            .expect("block exists");
+                        self.detach_block(batch, &index_lock_hashes, &block);
+                    }
+                    let index_state = LockHashIndexState {
+                        block_number: block.header().number() - 1,
+                        block_hash: block.header().parent_hash().to_owned(),
+                    };
+                    batch.insert_lock_hash_index_state(lock_hash, &index_state);
+                });
+            });
+
+        // attach blocks until reach tip or batch limit
+        let mut lock_hash_index_states = self.get_lock_hash_index_states();
+        let min_block_number: BlockNumber = lock_hash_index_states
+            .values()
+            .min_by_key(|index_state| index_state.block_number)
+            .expect("none empty index states")
+            .block_number;
+
+        // need to lock chain state, avoids inconsistent state in processing
+        let chain_state = self.shared.lock_chain_state();
```

Comment:
There is a way to avoid chain_state lock.
We can find the tip block first, then find it's ancestors until min_block_number.

Reply:
Since it is unnecessary to keep indexer 100% consistent with chain, I think it is worth to try to avoid the big `chain_state` lock, in case chain_state changed the indexer can resume the consistent in next sync.